ironfish/src/wallet/account/account.ts

```diff
@@ -251,6 +251,10 @@ export class Account {
       return
     }
 
+    if (owner.toString('hex') !== this.publicAddress) {
```

Comment:
I _think_ this is incorrect, and I'm surprised this doesn't have a failing test associated with it. If I'm not mistaken, the point of it syncing from the decrypted notes is because the wallet may want to fetch the info in the event of the account receiving a note of an asset, even if it isn't the owner.

As an example, I mint 5 Matcoin which creates a note of 5 Matcoin. Later in a different transaction, I send you 1 Matcoin, so the transaction will have 1 spend: my 5 Matcoin note, and 2 outputs/notes: 1. 1 Matcoin for you and 2. 4 Matcoin for me.

In this scenario, your wallet wants to fetch the asset info so it can display e.g. the asset name (Matcoin), even though it doesn't own the asset.

Reply:
Right, not sure how this escaped me, but I think this proves that our test coverage in this case is not as through as I thought. I will add more tests to catch this exact scenario

Reply:
Actually, this code as-is was correct, let me explain why. First of all, let me say that there is a test covering this scenario: it's `should save assets from the chain the account does not own`.

Your explanation is correct: when you receive a note from an unknown asset, that asset should be recorded in the wallet db, backfilling it from the chain db. This job is done by `upsertAssetsFromDecryptedNotes` which does the following (both now, in staging, and with this commit):

1. checks if the asset is in the wallet db
2. if it's not, gets the asset from the chain db
3. updates the asset on the wallet db with the latest block information

In staging, `upsertAssetsFromDecryptedNotes` writes to the wallet db twice: at step 2 and 3. With the original code from this PR, the wallet db is touched only at step 3.

I have updated this PR to re-add the write at step 2, but now that I dug deeper into the issue, I'm thinking of rolling back the change, as writing twice is unnecessary and a waste of resources. I see however how this can generate confusion, so I'll make sure that the code is properly documented.